feat(test runner): tags/annotations

dgozman opened the pull request:
API changes:
- `test(title, details, body)` where details contain `tag` and `annotation`.
- similar `details` property added to `test.skip`, `test.fail`, `test.fixme`, `test.only`, `test.describe` and other `test.describe.*` variations.
- `TestProject.tagFilter`/`TestConfig.tagFilter` that supports logical tag expressions with `(`, `)`, `and`, `or` and `not`.
- `--tag` CLI option to filter by tags.
- New annotations are available in `TestInfo.annotations` and `TestCase.annotations`.
- New tags are available in `TestCase.tags`.
    
Reporter changes:
- `json` reporter includes new tags in addition to old `@smoke`-style tags. **Breaking**: tags are now listed with the leading `@` symbol.
- `html` reporter filters by old and new tags with the same `@smoke` token.

Fixes #29229, fixes #23180.

mxschmitt commented on docs/src/test-api/class-test.md:
```diff
@@ -1552,6 +1552,73 @@ Step body.
 
 Whether to box the step in the report. Defaults to `false`. When the step is boxed, errors thrown from the step internals point to the step call site. See below for more details.
 
+## method: Test.tag
```
Should we add it on a describe as well?

mxschmitt commented on docs/src/test-api/class-testproject.md:
```diff
@@ -199,6 +199,34 @@ Use [`method: Test.describe.configure`] to change the number of retries for a sp
 Use [`property: TestConfig.retries`] to change this option for all projects.
 
+
+## property: TestProject.tagFilter
```
Should we add TestInfo.tags as well?

dgozman committed:
feat(test runner): tags/annotations

API changes:
- `test(title, details, body)` where details contain `tag` and `annotation`.
- similar `details` property added to `test.skip`, `test.fail`, `test.fixme`,
  `test.only`, `test.describe` and other `test.describe.*` variations.
- `TestProject.tagFilter`/`TestConfig.tagFilter` that supports logical tag expressions
  with `(`, `)`, `and`, `or` and `not`.
- `--tag` CLI option to filter by tags.
- New annotations are available in `TestInfo.annotations` and `TestCase.annotations`.
- New tags are available in `TestCase.tags`.

Reporter changes:
- `json` reporter includes new tags in addition to old `@smoke`-style tags.
  **Breaking**: tags are now listed with the leading `@` symbol.
- `html` reporter filters by old and new tags with the same `@smoke` token.
